File: lib/scripts/render-dashboard.js
#!/usr/bin/env node
"use strict";

const fs = require("fs");
const path = require("path");

function normalizeTitle(text) {
  return String(text || "")
    .toLowerCase()
    .replace(/[^a-z0-9]+/g, "-")
    .replace(/^-|-$/g, "");
}

function issueFilePath(issue) {
  if (issue.file) {
    return issue.file;
  }
  const fingerprint = issue.fingerprint || "";
  if (fingerprint && !/^[a-f0-9]{64}$/.test(fingerprint)) {
    const parts = fingerprint.split(":");
    if (parts.length >= 2) {
      return parts[1];
    }
  }
  return "";
}

function issueMatchKey(issue) {
  const repo = issue.repository || "";
  const file = issueFilePath(issue);
  if (repo && file) {
    return `${repo}|${file}`;
  }
  const title = normalizeTitle(issue.title);
  if (repo && title) {
    return `${repo}|${title}`;
  }
  return issue.id || "";
}

function deduplicateIssues(issues) {
  const groups = new Map();
  for (const issue of issues) {
    const key = issueMatchKey(issue);
    if (!groups.has(key)) {
      groups.set(key, []);
    }
    groups.get(key).push(issue);
  }

  const merged = [];
  for (const group of groups.values()) {
    if (group.length === 1) {
      merged.push(group[0]);
      continue;
    }
    const best = group
      .slice()
      .sort((a, b) => {
        const aIssue = String(a.id || "").startsWith("ISSUE-") ? 1 : 0;
        const bIssue = String(b.id || "").startsWith("ISSUE-") ? 1 : 0;
        if (bIssue !== aIssue) {
          return bIssue - aIssue;
        }
        return String(b.last_seen_at || "").localeCompare(String(a.last_seen_at || ""));
      })[0];
    merged.push(best);
  }
  return merged;
}

function loadJson(filePath, fallback) {
  try {
    return JSON.parse(fs.readFileSync(filePath, "utf8"));
  } catch {
    return fallback;
  }
}

function severityCounts(findings) {
  const counts = { High: 0, Medium: 0, Low: 0 };
  for (const finding of findings || []) {
    const severity = finding.severity || "Low";
    counts[severity] = (counts[severity] || 0) + 1;
  }
  return counts;
}

function rel(root, target) {
  if (!target) {
    return "";
  }
  const absolute = path.resolve(target);
  const relative = path.relative(root, absolute);
  if (!relative || relative.startsWith("..") || path.isAbsolute(relative)) {
    const value = String(target);
    if (value.startsWith("http://") || value.startsWith("https://")) {
      return value;
    }
    return value;
  }
  return relative.split(path.sep).join("/");
}

function main() {
  const root = path.resolve(process.argv[2] || process.env.LUMEN_WORKSPACE || process.cwd());
  const common = loadJson(path.join(root, "config", "common.json"), {});
  const repos = loadJson(path.join(root, "config", "repos.json"), { repositories: [] });
  const registry = loadJson(path.join(root, "state", "issue-registry.json"), { issues: [] });
  const product = common.product || {};
  const resultsDir = path.join(root, common.paths?.results_dir || "results");
  const logsDir = path.join(root, common.paths?.logs_dir || "logs");
  const dataPath = path.join(root, "dashboard-data.js");

  const runs = [];
  let latestRun = null;
  const latestPath = path.join(resultsDir, "scan-result.json");
  const latestData = loadJson(latestPath, {});
  if (Object.keys(latestData).length > 0) {
    latestRun = {
      id: path.basename(latestPath, ".json"),
      finished_at: latestData.finished_at || "",
      status: latestData.scan_status || "",
      findings: latestData.findings || [],
    };
  }

  const resultFiles = fs
    .readdirSync(resultsDir)
    .filter((name) => /^scan-result-.*\.json$/.test(name))
    .sort()
    .reverse();

  for (const name of resultFiles) {
    const filePath = path.join(resultsDir, name);
    const data = loadJson(filePath, {});
    if (!Object.keys(data).length) {
      continue;
    }
    const counts = severityCounts(data.findings);
    const report = data.report || {};
    const run = {
      id: path.basename(filePath, ".json"),
      started_at: data.started_at || "",
      finished_at: data.finished_at || "",
      status: data.scan_status || "",
      repos: data.repositories_scanned || 0,
      high: counts.High,
      medium: counts.Medium,
      low: counts.Low,
      prs: (data.prs || []).length,
      html: rel(root, report.html_path),
      pdf: rel(root, report.pdf_path),
      json: rel(root, filePath),
      feishu: data.feishu?.status || "",
      failures: (data.failures || []).length,
      findings: data.findings || [],
    };
    runs.push(run);
    if (!latestRun) {
      latestRun = {
        id: run.id,
        finished_at: run.finished_at,
        status: run.status,
        findings: run.findings,
      };
    }
  }

  const issues = deduplicateIssues(registry.issues || []);
  const issueCounts = {};
  for (const issue of issues) {
    const status = issue.status || "unknown";
    issueCounts[status] = (issueCounts[status] || 0) + 1;
  }

  const logs = fs
    .readdirSync(logsDir)
    .filter((name) => name.endsWith(".log"))
    .sort()
    .reverse()
    .slice(0, 10)
    .map((name) => ({
      name,
      href: rel(root, path.join(logsDir, name)),
    }));

  const payload = {
    generated_at: new Date().toISOString().replace(/\.\d{3}Z$/, "Z"),
    product: {
      name: product.name || "Lumen",
      tagline: product.tagline || "Illuminate code health across your repositories",
      codename: product.codename || "lumen",
    },
    config: common,
    repositories: repos.repositories || [],
    runs,
    latest_run: latestRun,
    issues,
    issue_counts: issueCounts,
    logs,
  };

  fs.writeFileSync(dataPath, `window.DASHBOARD_DATA = ${JSON.stringify(payload, null, 2)};\n`, "utf8");
  console.log(dataPath);
}

main();

File: lib/scripts/render-dashboard.py
#!/usr/bin/env python3
import json
import os
import re
import sys
from datetime import datetime, timezone
from pathlib import Path


def load_json(path: Path, default):
    try:
        with path.open(encoding="utf-8") as handle:
            return json.load(handle)
    except Exception:
        return default


def normalize_title(text) -> str:
    value = re.sub(r"[^a-z0-9]+", "-", str(text or "").lower())
    return value.strip("-")


def issue_file_path(issue: dict) -> str:
    if issue.get("file"):
        return issue["file"]
    fingerprint = issue.get("fingerprint") or ""
    if fingerprint and not re.fullmatch(r"[a-f0-9]{64}", fingerprint):
        parts = fingerprint.split(":")
        if len(parts) >= 2:
            return parts[1]
    return ""


def issue_match_key(issue: dict) -> str:
    repo = issue.get("repository") or ""
    file_path = issue_file_path(issue)
    if repo and file_path:
        return f"{repo}|{file_path}"
    title = normalize_title(issue.get("title"))
    if repo and title:
        return f"{repo}|{title}"
    return issue.get("id") or ""


def deduplicate_issues(issues: list) -> list:
    groups = {}
    for issue in issues:
        key = issue_match_key(issue)
        groups.setdefault(key, []).append(issue)

    merged = []
    for group in groups.values():
        if len(group) == 1:
            merged.append(group[0])
            continue
        best = sorted(
            group,
            key=lambda item: (
                1 if str(item.get("id", "")).startswith("ISSUE-") else 0,
                str(item.get("last_seen_at") or ""),
            ),
            reverse=True,
        )[0]
        merged.append(best)
    return merged


def severity_counts(findings):
    counts = {"High": 0, "Medium": 0, "Low": 0}
    for finding in findings or []:
        severity = finding.get("severity", "Low")
        counts[severity] = counts.get(severity, 0) + 1
    return counts


def rel(root: Path, path) -> str:
    if not path:
        return ""
    try:
        return str(Path(path).resolve().relative_to(root.resolve()))
    except Exception:
        value = str(path)
        if value.startswith("http://") or value.startswith("https://"):
            return value
        return value


def main() -> int:
    workspace_arg = sys.argv[1] if len(sys.argv) > 1 else os.environ.get("LUMEN_WORKSPACE", ".")
    root = Path(workspace_arg).resolve()
    common = load_json(root / "config" / "common.json", {})
    repos = load_json(root / "config" / "repos.json", {"repositories": []})
    registry = load_json(root / "state" / "issue-registry.json", {"issues": []})
    product = common.get("product", {})
    results_dir = root / common.get("paths", {}).get("results_dir", "results")
    logs_dir = root / common.get("paths", {}).get("logs_dir", "logs")
    data_path = root / "dashboard-data.js"

    runs = []
    latest_run = None
    latest_path = results_dir / "scan-result.json"
    latest_data = load_json(latest_path, {})
    if latest_data:
        latest_run = {
            "id": latest_path.stem,
            "finished_at": latest_data.get("finished_at", ""),
            "status": latest_data.get("scan_status", ""),
            "findings": latest_data.get("findings", []),
        }

    for path in sorted(results_dir.glob("scan-result-*.json"), reverse=True):
        data = load_json(path, {})
        if not data:
            continue
        counts = severity_counts(data.get("findings", []))
        report = data.get("report", {})
        run = {
            "id": path.stem,
            "started_at": data.get("started_at", ""),
            "finished_at": data.get("finished_at", ""),
            "status": data.get("scan_status", ""),
            "repos": data.get("repositories_scanned", 0),
            "high": counts["High"],
            "medium": counts["Medium"],
            "low": counts["Low"],
            "prs": len(data.get("prs", [])),
            "html": rel(root, report.get("html_path")),
            "pdf": rel(root, report.get("pdf_path")),
            "json": rel(root, path),
            "feishu": data.get("feishu", {}).get("status", ""),
            "failures": len(data.get("failures", [])),
            "findings": data.get("findings", []),
        }
        runs.append(run)
        if latest_run is None:
            latest_run = {
                "id": run["id"],
                "finished_at": run["finished_at"],
                "status": run["status"],
                "findings": run["findings"],
            }

    issues = deduplicate_issues(registry.get("issues", []))
    issue_counts = {}
    for issue in issues:
        status = issue.get("status", "unknown")
        issue_counts[status] = issue_counts.get(status, 0) + 1

    logs = [
        {"name": path.name, "href": rel(root, path)}
        for path in sorted(logs_dir.glob("*.log"), reverse=True)[:10]
    ]

    payload = {
        "generated_at": datetime.now(timezone.utc).replace(microsecond=0).isoformat().replace("+00:00", "Z"),
        "product": {
            "name": product.get("name", "Lumen"),
            "tagline": product.get("tagline", "Illuminate code health across your repositories"),
            "codename": product.get("codename", "lumen"),
        },
        "config": common,
        "repositories": repos.get("repositories", []),
        "runs": runs,
        "latest_run": latest_run,
        "issues": issues,
        "issue_counts": issue_counts,
        "logs": logs,
    }

    js_text = "window.DASHBOARD_DATA = " + json.dumps(payload, ensure_ascii=False, indent=2) + ";\n"
    data_path.write_text(js_text, encoding="utf-8")
    print(data_path)
    return 0


if __name__ == "__main__":
    raise SystemExit(main())
